lib/ace/ext/papeeria/tex_completer.coffee

```diff
@@ -84,16 +92,73 @@ define( (require, exports, module) ->
     value: word
     meta: 'equation'
   )
+  istype = (token, type) -> 
+    return token.type.lastIndexOf(type) > -1
+  
+
+  #we will move this code into tex_completer in next pull request and well make this beautiful
+  init = (editor, bindKey) -> 
+    HashHandler = require("ace/keyboard/hash_handler").HashHandler; 
+    keyboardHandler = new HashHandler();
+    keyboardHandler.addCommand(
+      name: 'add item in list mode',
+      bindKey: bindKey,
+      exec: (editor) ->  
+        pos = editor.getCursorPosition()
+        curLine = editor.session.getLine(pos.row);
+        intendCount = ContextHelper.getNestingOfList(editor.session, pos.row)
+        # it's temporary fix bug with added \item before \begin{itemize|enumerate}
+        if ContextHelper.getContext(editor.session, pos.row) == LIST_STATE && curLine.indexOf("begin") < pos.column
+          editor.insert("\n" + "    ".repeat(intendCount) + "\\item ")
+          return true
+        else 
+          return false
+        
+    );
+    editor.keyBinding.addKeyboardHandler(keyboardHandler);
+
+  class ReferenceGetter
+    constructor: ->
+      @cachedURL =  ""
+      @cache = []
+    getReference: (url, callback) -> 
+      if (url == @cashedURL)
+        return @cashe
+      else 
+        @cachedURL = url
+        json = $.getJSON(url)
+        thisRef = @
+        json.success((data) ->
+          callback(data, thisRef)
+        )
+
+        return @cache
 
-  exports.getCompletions = (editor, session, pos, prefix, callback) ->
-    context = session.getContext(pos.row)
-    if context == "start"
-      callback(null, listSnippets.concat(equationSnippets.concat(basicSnippets)))
+  exports.CompletionTools = 
+    TexCompleter: class TexCompleter
+      constructor: ->
+        @r = new ReferenceGetter()
+      getCompletions: (editor, session, pos, prefix, callback) ->
+        context = ContextHelper.getContext(session, pos.row)
+        token = session.getTokenAt(pos.row, pos.column)
+        console.log(token)
+        if istype(token, "ref")
+          callback(null, @r.getReference("example.json", (data, r) -> 
+            r.cache = data.map((elem) -> 
+                            return {
```

Comment:
too many spaces. For contonuations (statement wrapped to the next line) it is okay to indent with 4 spaces.